jni/include/native_engines_stream_support.h

```diff
@@ -119,6 +121,110 @@ class NativeEngineIndexInputMediator {
 
 
 
+/**
+ * This class delegates the provided index output to do IO processing.
+ * In most cases, it is expected that IndexOutputWithBuffer was passed down to this,
+ * which eventually have Lucene's IndexOutput to write bytes.
+ */
+class NativeEngineIndexOutputMediator {
+ public:
+  NativeEngineIndexOutputMediator(JNIUtilInterface *_jni_interface,
+                                  JNIEnv *_env,
+                                  jobject _indexOutput)
+      : jni_interface(_jni_interface),
+        env(_env),
+        indexOutput(_indexOutput),
+        bufferArray((jbyteArray) (_jni_interface->GetObjectField(_env,
+                                                                 _indexOutput,
+                                                                 getBufferFieldId(_jni_interface, _env)))),
+        writeBytesMethod(getWriteBytesMethod(_jni_interface, _env)),
+        bufferLength(jni_interface->GetJavaBytesArrayLength(env, bufferArray)),
+        nextWriteIndex() {
+  }
+
+  void writeBytes(uint8_t *source, size_t nbytes) {
+    auto left = nbytes;
+    while (left > 0) {
+      const auto writeBytes = std::min(bufferLength - nextWriteIndex, left);
+
+      // === Critical Section Start ===
+
+      // Get primitive array pointer, no copy is happening in OpenJDK.
+      auto primitiveArray =
+          (jbyte *) jni_interface->GetPrimitiveArrayCritical(env, bufferArray, nullptr);
+
+      // Copy the given bytes to Java byte[] address.
+      std::memcpy(primitiveArray + nextWriteIndex, source, writeBytes);
+
+      // Release the acquired primitive array pointer.
+      // 0 tells JVM to copy back the content, and to free the pointer. It will be ignored if we acquired an internal
+      // primitive array pointer instead of a copied version.
+      // From JNI docs:
+      // Mode 0 : copy back the content and free the elems buffer
+      // The mode argument provides information on how the array buffer should be released. mode has no effect if elems
+      // is not a copy of the elements in array.
+      jni_interface->ReleasePrimitiveArrayCritical(env, bufferArray, primitiveArray, 0);
```

Comment:
Could we use  RAII (Resource Acquisition Is Initialization) to manage critical sections helps to ensure that resources are released even if an exception occurs. We can create a helper class to manage the JNI array critical section safely. any thought?

Reply:
No exceptions will be thrown by design!
all the methods in the critical sections are marked with `nothrow`.

Reply:
Make sense , if there is no expection thrown we indeed don't need RAII , but can gurantee that this code will not be extended in future and no expection ever be thrown ..I was just trying to be cautious here !!!